functions.sql

```diff
@@ -0,0 +1,92 @@
+CREATE OR REPLACE FUNCTION carto_int_access(primary_mode text, accesstag text)
+	RETURNS text
+	LANGUAGE SQL
+	IMMUTABLE PARALLEL SAFE
+AS $$
+SELECT
+	CASE
+	WHEN accesstag IN ('yes', 'designated', 'permissive', 'customers') THEN 'yes'
+	WHEN accesstag IN ('destination',  'delivery', 'permit') THEN
+		CASE WHEN primary_mode IN ('motorcar', 'pedestrian') THEN 'restricted' ELSE 'no' END
+	WHEN accesstag IN ('no', 'private', 'agricultural', 'forestry', 'agricultural;forestry') THEN 'no'
+	ELSE NULL
+	END
+END
+$$;
+
+/* Try to promote path to cycleway (if bicycle allowed), then bridleway (if horse)
+   This duplicates existing behaviour where designated access is required */
+CREATE OR REPLACE FUNCTION carto_path_primary_mode(bicycle text, horse text)
+	RETURNS text
+	LANGUAGE SQL
+	IMMUTABLE PARALLEL SAFE
+AS $$
+SELECT
+	CASE
+		WHEN bicycle IN ('designated') THEN 'bicycle'
+		WHEN horse IN ('designated') THEN 'horse'
+		ELSE 'foot'
+	END
+END
+$$;
+
+/* Classify highways into access categories which will be treated in the same way
+   Default is NULL in which case only the access tag will be used (e.g. highway=track)
+   Note that bicycle, horse arguments are only relevant if considering highway=path */
+CREATE OR REPLACE FUNCTION carto_highway_primary_mode(highway text, bicycle text, horse text)
+	RETURNS text
+	LANGUAGE SQL
+	IMMUTABLE PARALLEL SAFE
+AS $$
+SELECT
+	CASE 
+    WHEN highway IN ('motorway', 'motorway_link', 'trunk', 'trunk_link', 'primary', 'primary_link', 'secondary',
+                 'secondary_link', 'tertiary', 'tertiary_link', 'residential', 'unclassified', 'living_street', 'service', 'road') THEN 'motorcar'
+	WHEN highway = 'pedestrian' THEN 'pedestrian'
+	WHEN highway IN ('footway', 'steps') THEN 'foot'
+	WHEN highway = 'bridleway' THEN 'horse'
+	WHEN highway = 'cycleway' THEN 'bicycle'
+	WHEN highway = 'path' THEN carto_path_primary_mode(bicycle, horse)
+	ELSE NULL
+	END
+END
+$$;
+
+/* Return int_access value which will be used to determine access marking.
+   Only a restricted number of types can be returned, with NULL corresponding to no access restriction */
+CREATE OR REPLACE FUNCTION carto_highway_int_access(highway text, "access" text, foot text, bicycle text, horse text, motorcar text, motor_vehicle text, vehicle text)
+  RETURNS text
+  LANGUAGE SQL
+  IMMUTABLE PARALLEL SAFE
+AS $$
+SELECT
+	CASE carto_highway_primary_mode(highway, bicycle, horse)
+	WHEN 'motorcar' THEN
+		carto_int_access('motorcar', CASE
+			WHEN motorcar IN ('yes', 'designated', 'permissive', 'no', 'private', 'destination', 'customers', 'delivery', 'permit') THEN motorcar
+			WHEN motor_vehicle IN ('yes', 'designated', 'permissive', 'no', 'private', 'destination', 'customers', 'delivery', 'permit') THEN motor_vehicle
+			WHEN vehicle IN ('yes', 'designated', 'permissive', 'no', 'private', 'destination', 'customers', 'delivery', 'permit') THEN vehicle
+			ELSE "access" END)
+	WHEN 'pedestrian' THEN carto_int_access('pedestrian', 'no')
```

Comment:
This is odd - if you have a distinct primary mode `pedestrian` (and it is unclear what this means relative to `foot`): why does this universally mean `access=no`.  Note that while we currently do not render `highway=pedestrian` with access dashing, access restricted pedestrian roads of course do exist - like in private parks or in military facilities.

Reply:
Again, perhaps renaming "access type" as "primary mode" has hindered rather than helped. The idea of a "pedestrian" access mode was an implicit `motorcar/motor_vehicle=no`. Putting this `no` into `int_access` simplifies later queries about whether to add colour for `bus=yes` etc.; only ways with `int_access=restricted|no` are considered.
The alternative is to set `int_access=NULL` and to treat `highway=pedestrian` as a special case [`IF int_access=restricted|no OR highway=pedestrian THEN <consider bus=yes marking>`]